cmd/kube-proxy/app/server.go

```diff
@@ -130,7 +131,25 @@ type Options struct {
 }
 
 // AddFlags adds flags to fs and binds them to options.
-func (o *Options) AddFlags(fs *pflag.FlagSet) {
+func (o *Options) AddFlags(mainfs *pflag.FlagSet) {
+	fs := pflag.NewFlagSet("", pflag.ExitOnError)
+	defer func() {
+		// All KubeProxyConfiguration flags are now deprecated
+		deprecated := "This parameter should be set via the config file specified by the Kube-proxy's --config flag."
+		notDeprecated := map[string]bool{
+			"config":          true,
+			"write-config-to": true,
+			"cleanup":         true,
+		}
+		fs.VisitAll(func(f *pflag.Flag) {
+			if notDeprecated[f.Name] {
+				return
+			}
+			f.Deprecated = deprecated
+		})
+		mainfs.AddFlagSet(fs)
```

Comment:
So now we are back at the point where new flags gets added by the PR (`--logging-format`, etc.) which trigger a deprecation warning:

```
$ go run ./cmd/kube-proxy/ --logging-format=json
Flag --logging-format has been deprecated, This parameter should be set via the config file specified by the Kube-proxy's --config flag.
{"ts":1684218173834.7542,"caller":"app/server_others.go:75","msg":"Using iptables proxy","v":0}
{"ts":1684218173834.9937,"caller":"app/server.go:554","msg":"Neither kubeconfig file nor master URL was specified, falling back to in-cluster config","v":0}
{"ts":1684218173835.0076,"caller":"app/server.go:507","msg":"Error running ProxyServer","err":"unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined"}
{"ts":1684218173835.024,"caller":"cli/run.go:74","msg":"command failed","err":"unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined"}
exit status 1
```

I think we have two options:
1. Not add the logging flags to kube-proxy, only configure logging through the config file.
2. Add the flags without marking them as deprecated.

I prefer option 2 because it's consistent with other commands, but I would also be okay with option 1.

Reply:
I also think the option2 is better, but I see kubelet also marked logging flag as deprecated, it's bug here?

```shell
➜  kubernetes git:(loggin_format) go run ./cmd/kubelet/kubelet.go --logging-format json 
Flag --logging-format has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.
{"ts":1684218864145.937,"caller":"cli/run.go:74","msg":"command failed","err":"failed to construct kubelet dependencies: error reading /var/lib/kubelet/pki/kubelet.key, certificate and key must be supplied as a pair"}
exit status 1

```

Reply:
I suppose it wasn't discussed and/or considered at that time.

Reply:
So kubelet also needs to be updated?

Reply:
Let's first reach consensus on which of the two options are preferred.

This might be a topic for SIG Architecture or SIG CLI.

@liggitt ?

Reply:
for components which support a config file that has ~parity with all their CLI flags AND has forward-compatibility guarantees (beta or stable), the flags have generally been marked deprecated in favor of the config-file-specified options

so far, I think that's just kube-scheduler and kubelet

I would not expect kube-proxy to mark CLI flags as deprecated in favor of an alpha-level config file

Reply:
yeah, so we don't mark CLI flags as deprecated, Just warn the user that the flag will be deprecated and use the config file instead? Just like the current PR changes, Do you like it？

Reply:
we shouldn't warn about use of flags in favor of alpha-level config files, only once those graduate to versions with forward compatibility guarantees

Reply:
thanks, updated.